### PR TITLE
Feat add bitcoin anchor validation

### PR DESCRIPTION
**Pull request by mr-onadasky**
Add validation for bitcoin anchors

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,11 +12,14 @@ npm install --save @settlemint/certimint-validate
...
 
-    import { CertiMintValidation } from '@settlemint/certimint-validate';
+import { CertiMintValidation } from '@settlemint/certimint-validate';
 
-    const certiMintValidation = new CertiMintValidation();
-    isValid = await certiMintValidation.validateSeal(
-      mySealObject,
-      'https://mainnet.infura.io'
-    );
+const certiMintValidation = new CertiMintValidation(
```
Comment (peterjan) [inline README.md]: PP: the formatting of this readme could definitely be improved :)

File: src/validate.ts [hunk 2]
```diff
@@ -159,25 +160,44 @@ export class CertiMintValidation {
             anchor.merkleRoot
           );
 
-          isValid = isValid && anchor.exists && validProof;
+          isValid = anchor.exists && validProof;
```
Comment (peterjan) [inline src/validate.ts]: REQ: you broke the logic here, you allow for isValid to go from false -> true if it's followed by a valid proof and/or anchor. The `isValid = isValid &&` trick basically ensures that once invalid you can never go back to valid :) it was there for a reason

File: src/validate.ts [hunk 3]
```diff
@@ -159,25 +160,44 @@ export class CertiMintValidation {
...
+
+            isValid = anchor.exists && validProof;
           } catch (error) {
-            console.log(error);
+            if (error.response !== undefined && error.response.status === 429) {
+              throw new Error(
+                'Too many request to the blockcypher api, please add an apikey or upgrade your blockcypher plan'
+              );
+            } else {
+              throw error;
```
Comment (peterjan) [inline src/validate.ts]: REQ: watch out - you changed logic here again - we used to fail and silently and log. Unsure what to do here, depends on who's calling and how that's being handled. I think yours is better but we need to be sure.

**Review by peterjan — CHANGES_REQUESTED** [review]
Not too bad - fix those logic mistakes and we can merge.
I wish you had reverted that mistakenly committed commit though.
You can never do wrong with `git revert [commitsha]` on your own mistakes.
But let's leave as is now and get this in asap.

File: src/validate.ts (hunk 3 above)
Comment (mr-onadasky) [reply src/validate.ts]: Q: So should I remove this code then or should I rewrite all the error catching in general from this repo?

File: src/validate.ts [hunk 4]
```diff
@@ -133,63 +138,110 @@ export class CertiMintValidation {
...
 
-          anchor.exists = tx.data === this.addHexPrefix(anchor.merkleRoot);
+    return isValid;
+  }
 
-          const merkleTools = new MerkleTools({
-            hashType: 'SHA3-512'
-          });
+  private async validateEthereumAnchor(
+    anchors: any,
```
Comment (peterjan) [inline src/validate.ts]: REQ: please type this

File: src/validate.ts [hunk 5]
```diff
@@ -133,63 +138,110 @@ export class CertiMintValidation {
...
+        dataHash,
+        anchor.merkleRoot
+      );
+
+      return isValid && anchor.exists && validProof;
+    }
+  }
+
+  private async validateBitcoinAnchor(
+    anchors: any,
```
Comment (peterjan) [inline src/validate.ts]: REQ: please type this

**Review by peterjan — CHANGES_REQUESTED** [review]
final changes, looks good

File: src/validate.ts [hunk 6]
```diff
@@ -133,63 +138,110 @@ export class CertiMintValidation {
...
-      if (protocol === Protocol.ETHEREUM) {
-        for (const chainId of Object.keys(anchors[protocol])) {
-          const anchor = anchors[protocol][chainId];
-          const provider = new JsonRpcProvider(anchor.nodeUrl);
-
-          const tx = await provider.getTransaction(
-            this.addHexPrefix(anchor.transactionId)
+      switch (protocol) {
+        case Protocol.ETHEREUM:
+          isValid = await this.validateEthereumAnchor(
```
Comment (peterjan) [inline src/validate.ts]: REQ: this is wrong - it needs to be `isValid = isValid && ...`

File: src/validate.ts [hunk 7]
```diff
@@ -133,63 +138,110 @@ export class CertiMintValidation {
...
+            anchors,
+            dataHash,
+            isValid
+          );
+          break;
+        case Protocol.BITCOIN:
+          isValid = await this.validateBitcoinAnchor(
```
Comment (peterjan) [inline src/validate.ts]: REQ: this is wrong - it needs to be `isValid = isValid && ...`

**Review by peterjan — APPROVED** [review]

**Pull request merged**